File: [O]math/[O]math/338-counting-bits.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        res = [0 for _ in range(n + 1)]
        for i in range(n + 1):
            if i % 2 == 0:
                res[i] = res[i // 2]
            else:
                res[i] = res[i // 2] + 1
        return res
    
# time O(n)
# space O(n)
# using math

File: [O]math/[O]math/338-counting-bits.java
class Solution {
    public int[] countBits(int n) {
        int[] res = new int[n + 1];
        for (int i = 0; i < n + 1; i++) {
            if (i % 2 == 1) {
                res[i] = res[i / 2] + 1;
            } else {
                res[i] = res[i / 2];
            }
        }
        return res;
    }
}

// time O(n)
// space O(n)
// using math